lib/src/repo_path.rs

```diff
@@ -468,6 +468,77 @@ fn is_valid_repo_path_str(value: &str) -> bool {
     !value.starts_with('/') && !value.ends_with('/') && !value.contains("//")
 }
 
+/// An error from `RepoPathUiConverter::parse_file_path`.
+#[derive(Debug, Error)]
+#[error(r#"Invalid path "{input}""#)]
+pub struct UiPathParseError {
+    /// Input path.
+    pub input: String,
+    /// Source error.
+    pub source: Box<dyn std::error::Error + Send + Sync>,
+}
+
+/// Converts `RepoPath`s to and from plain strings as displayed to the user
+/// (e.g. relative to CWD).
+pub trait RepoPathUiConverter: Debug {
+    /// Format a path for display in the UI.
+    fn format_file_path(&self, file: &RepoPath) -> String;
+
+    /// Parses a path from the UI.
+    ///
+    /// It's up to the implementation whether absolute paths are allowed, and
+    /// where relative paths are interpreted as relative to.
+    fn parse_file_path(&self, input: &str) -> Result<RepoPathBuf, UiPathParseError>;
+}
```

Comment:
Out of curiosity, do we need trait extensibility for path converter?

I thought it would be just "cwd-relative path" or "noop" (for server-side use case.) The latter can be modeled as `cwd: "", base: ""` (or `enum Fs(FsRepoPathUiConverter)|Noop` if we want type safety there.)

Reply:
I think it's useful for the server implementation to be independent of `Path`, so it can parse and format paths consistently regardless of which platform it's running on (e.g. using forward slashes even if the server is running on Windows).

Reply:
If there aren't many variants and the implementations are known, I would use `enum` until it become a mess. Trait object adds various restrictions such as it can't be cloned by default.

`enum Fs|Noop` might allow us to reject explicit `cwd:<path>` prefix (which doesn't make sense at server side) if needed.